crates/integration-tests/src/federation/websocket_request.rs

```diff
@@ -0,0 +1,69 @@
+//! graphql-ws-client glue code
+
+use std::future::IntoFuture;
+
+use futures::{future::BoxFuture, stream::BoxStream};
+use url::Url;
+
+use super::{GraphQlRequest, GraphqlResponse};
+
+pub struct WebsocketRequest {
+    pub(super) request: GraphQlRequest,
+    pub(super) init_payload: Option<serde_json::Value>,
+    pub(super) router: axum::Router<()>,
+}
+
+impl IntoFuture for WebsocketRequest {
+    type Output = Result<BoxStream<'static, GraphqlResponse>, graphql_ws_client::Error>;
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        use async_tungstenite::tungstenite::{client::IntoClientRequest, http::HeaderValue};
+        use futures_util::StreamExt;
+
+        Box::pin(async move {
+            let handler = self.router.into_make_service();
+            let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
+
+            let mut url: Url = format!("http://{}", listener.local_addr().unwrap()).parse().unwrap();
+            url.set_path("/ws");
+            url.set_scheme("ws").unwrap();
+
+            // FIXME: we should shut down the server when the stream is dropped.
+            tokio::spawn(axum::serve(listener, handler).into_future());
+
+            let mut request = url.as_ref().into_client_request().unwrap();
+
+            request.headers_mut().insert(
+                "Sec-WebSocket-Protocol",
+                HeaderValue::from_str("graphql-transport-ws").unwrap(),
+            );
+
+            let (connection, _) = async_tungstenite::tokio::connect_async(request).await.unwrap();
+
+            let (client, actor) = graphql_ws_client::Client::build(connection)
+                .payload(self.init_payload.unwrap_or_else(|| serde_json::Map::default().into()))?
```

Comment:
shouldn't the default be `null`?

Reply:
Yeah, can be null (https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#connectioninit), will change to null when I deal with the fixmes, before merging.